util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+  }
+
+  virtual ~WindowsMmapReadableFile() {
+    ::UnmapViewOfFile(mmapped_region_);
+    CloseHandle(filemap_);
+    limiter_->Release();
+  }
+
+  virtual Status Read(uint64_t offset, size_t n, Slice* result,
+                      char* scratch) const {
+    Status s;
+    if (offset + n > length_) {
+      *result = Slice();
+      s = WindowsError(filename_, ERROR_INVALID_PARAMETER);
+    } else {
+      *result = Slice(reinterpret_cast<char*>(mmapped_region_) + offset, n);
+    }
+    return s;
+  }
+};
+
+
+class WindowsWritableFile : public WritableFile {
+ private:
+  std::string filename_;
+  HANDLE handle_;
+
+ public:
+  WindowsWritableFile(const std::string& fname, HANDLE handle)
+      : filename_(fname), handle_(handle) { }
+
+  virtual ~WindowsWritableFile() {
+    if (handle_ != INVALID_HANDLE_VALUE) {
+      // Ignoring any potential errors
+      CloseHandle(handle_);
+    }
+  }
+
+  virtual Status Append(const Slice& data) {
+    DWORD wrote = 0;
+    if (!::WriteFile(handle_, data.data(), data.size(), &wrote, NULL)) {
+      return WindowsError(filename_, GetLastError());
+    }
+    if (wrote != data.size()) {
+      return Status::IOError(filename_, "Could not write entire buffer to disk");
+    }
+    return Status::OK();
+  }
+
+  virtual Status Close() {
+    Status result;
+    if (!::CloseHandle(handle_)) {
+      result = WindowsError(filename_, GetLastError());
+    }
+    handle_ = INVALID_HANDLE_VALUE;
+    return result;
+  }
+
+  virtual Status Flush() {
+    if (!::FlushFileBuffers(handle_)) {
+      return WindowsError(filename_, GetLastError());
+    }
+    return Status::OK();
+  }
+
+  Status SyncDirIfManifest() {
+    // Windows does not have the concept of flushing a diectory.
+    // Although you can create a HANDLE to a directory with
+    // CreateFile using FILE_FLAG_BACKUP_SEMANTICS, you cannot
+    // call FileFlushBuffers() or any other data syncing routine.
+    return Status::OK();
+  }
+
+  virtual Status Sync() {
+    // Ensure new files referred to by the manifest are in the filesystem.
+    Status s = SyncDirIfManifest();
+    if (!s.ok()) {
+      return s;
+    }
+    return Flush();
+  }
+};
+
+
+static std::streamsize file_length(HANDLE handle)
+{
+  LARGE_INTEGER bytes;
+  if (!::GetFileSizeEx(handle, &bytes)) {
+      throw WindowsError("", GetLastError());
+  }
+  return static_cast<std::streamsize>(bytes.QuadPart);
+}
+
+
+static int LockOrUnlock(HANDLE handle, bool lock)
+{
+  LARGE_INTEGER bytes;
+  bytes.QuadPart = file_length(handle);
+  if (lock) {
+    return ::LockFile(handle, 0, 0, bytes.LowPart, bytes.HighPart) ? 0 : -1;
+  } else {
+    return ::UnlockFile(handle, 0, 0, bytes.LowPart, bytes.HighPart) ? 0 : -1;
+  }
+}
+
+
+class WindowsFileLock : public FileLock {
+ public:
+  HANDLE handle_;
+  std::string name_;
+};
+
+// Set of locked files.  We keep a separate set instead of just
+// relying on fcntrl(F_SETLK) since fcntl(F_SETLK) does not provide
```

Comment:
fcntl comments seem stale.

Also, is this lock table really needed on Windows?  I am not sure if LockFile/UnlockFile have the same problem as fcntl(F_SETLK) does.